### PR TITLE
PromQL: Use PositionRange in parser errors

### PR DESCRIPTION
**Pull request by slrtbtfs**
This PR changes the `ParseErr` struct of the Prometheus parser to use the new `PositionRange` structure. This is especially useful for the language server, since it allows underlining the wrong parts in expressions with errors.

Additionally type errors now report accurate positions. Before, all type errors had the end of the query as their position.

### CODE REVIEW

File: promql/parse.go [hunk 1]
```diff
@@ -151,19 +175,16 @@ func (p *parser) typecheck(node Node) (err error) {
 }
 
 // errorf formats the error and terminates processing.
-func (p *parser) errorf(format string, args ...interface{}) {
-	p.error(errors.Errorf(format, args...))
+func (p *parser) errorf(positionRange PositionRange, format string, args ...interface{}) {
```
Comment (brian-brazil) [inline promql/parse.go]: Having an errorf that doesn't have the usual printf/errorf signature is weird.

File: promql/parse.go [hunk 2]
```diff
@@ -281,12 +302,23 @@ func (p *parser) newBinaryExpression(lhs Node, op Item, modifiers Node, rhs Node
 	ret.RHS = rhs.(Expr)
 	ret.Op = op.Typ
 
+	// opRange returns the PositionRange of the operator part of the BinaryExpr.
+	// This is made a function instead of a variable, so it is lazily evaluated on demand.
+	opRange := func() (r PositionRange) {
+		//Remove whitespace at the beginning and end of the range
```
Comment (brian-brazil) [inline promql/parse.go]: Formatting.

File: promql/parse.go (hunk 1 above)
Comment (slrtbtfs) [reply promql/parse.go]: So I should rename it?

File: promql/parse.go (hunk 2 above)
Comment (slrtbtfs) [reply promql/parse.go]: What exactly is wrong with formatting?

The unnecessary line breaks in the for loops are enforced by gofmt.

Comment (brian-brazil) [reply promql/parse.go]: No space after the //, no .

File: promql/parse.go (hunk 1 above)
Comment (brian-brazil) [reply promql/parse.go]: That'd be best I think

File: promql/parse.go (hunk 2 above)
Comment (slrtbtfs) [reply promql/parse.go]: Oh, I see.

File: promql/parse.go [hunk 3]
```diff
@@ -151,19 +175,16 @@ func (p *parser) typecheck(node Node) (err error) {
 }
 
 // errorf formats the error and terminates processing.
```
Comment (brian-brazil) [inline promql/parse.go]: Comments needs updating.

**Comment by roidelapluie** [comment]
What about errorfAtPos ?

**Comment by slrtbtfs** [comment]
I'm working on another PR which will allow having a list of errors.

`addError` and `addErrorf` might be an option, then.

**Comment by brian-brazil** [comment]
Thanks!

**Comment by roidelapluie** [comment]
I have to say that the result is quite impressive. Thanks!

**Pull request merged**